File: php/25-reverse-nodes-in-k-group.php
<?php
/**
 * Definition for a singly-linked list.
 * class ListNode {
 *     public $val = 0;
 *     public $next = null;
 *     function __construct($val) { $this->val = $val; }
 * }
 */
class Solution 
{
    /**
     * @param ListNode $head
     * @param Integer $k
     * @return ListNode
     */
    public function reverseKGroup($head, $k) 
	{
        if ($k == 0) {
            return $head;
        }
        $stack = [];
        $dummy = new ListNode(0);
        $node = $dummy;
        while ($head != null) {
            $next = $head->next;
            $head->next = null;
            array_push($stack, $head);
            if (count($stack) == $k) {
                while ($stack) {
                    $node->next = array_pop($stack);                    
                    $node = $node->next;
                }
            }
            $head = $next;
        }
        if (!empty($stack)) {
            $count = count($stack);
            for ($i = 0; $i < $count; $i++) {
                $node->next = $stack[$i];                    
                $node = $node->next;
            }
        }
        return $dummy->next;
    }
}

File: go/25-reverse-nodes-in-kgroup.go
package main

/**
 * Definition for singly-linked list.
 * type ListNode struct {
 *     Val int
 *     Next *ListNode
 * }
 */
func reverseKGroup(head *ListNode, k int) *ListNode {
	if head == nil || k == 0 {
		return head
	}
	dummy := &ListNode{}
	node := dummy
	stack := []*ListNode{}

	for head != nil {
		next := head.Next
		head.Next = nil
		stack = append(stack, head)
		if len(stack) == k {
			for len(stack) > 0 {
				l := len(stack)
				node.Next = stack[l-1]
				stack = stack[0 : l-1]
				node = node.Next
			}
		}
		head = next
	}
	for len(stack) > 0 {
		node.Next = stack[0]
		node = node.Next
		stack = stack[1:]
	}
	return dummy.Next
}
